test/python/transpiler/test_ross_selinger.py

```diff
@@ -0,0 +1,174 @@
+from qiskit.transpiler.passes.synthesis import RossSelingerSynthesis
+
+from test import QiskitTestCase  # pylint: disable=wrong-import-order
+
+
+# Set of single-qubit Clifford gates
+CLIFFORD_GATES_1Q_SET = {"id", "x", "y", "z", "h", "s", "sdg", "sx", "sxdg"}
+
+# Set of Clifford+T gates
+CLIFFORD_T_GATES_SET = set(get_clifford_gate_names() + ["t", "tdg"])
+
+
+@ddt
+class TestRossSelingerSynthesis(QiskitTestCase):
+    """Test Ross-Selinger synthesis methods."""
+
+    def test_gridsynth_rz_correct(self):
+        """Test that gridsynth_rz works correctly."""
+        num_trials = 40
+        for angle in np.linspace(-2 * np.pi, 2 * np.pi, num_trials):
+            with self.subTest(angle=angle):
+                # Approximate RZ-rotation
+                approximate_circuit = gridsynth_rz(angle, 1e-10)
+                # Check the operators are (almost) equal
+                self.assertEqual(Operator(approximate_circuit), Operator(RZGate(angle)))
+
+    @data(10, -10)
+    def test_gridsynth_rz_with_nonstandard_angles(self, angle):
+        """Test that gridsynth_rz works correctly."""
+        # Approximate RZ-rotation
+        approximate_circuit = gridsynth_rz(angle, 1e-10)
+        # Check the operators are (almost) equal
+        self.assertEqual(Operator(approximate_circuit), Operator(RZGate(angle)))
+
+    def test_gridsynth_unitary_correct(self):
+        """Test that gridsynth_unitary works correctly."""
+        num_trials = 50
+        for seed in range(num_trials):
+            with self.subTest(seed=seed):
+                # Create a random 1q unitary.
+                unitary = random_unitary(2, seed=seed)
+                # Approximate unitary
+                approximate_circuit = gridsynth_unitary(unitary.data, 1e-10)
+                # Check the operators are (almost) equal
+                self.assertEqual(Operator(approximate_circuit), Operator(unitary))
+
+    def test_gridsynth_rz_deterministic(self):
+        """Test that calling gridsynth_rz multiple times produces the same circuit."""
+        angle = 1.2345
+        num_trials = 10
+        approximate_circuits = [gridsynth_rz(angle, 1e-10) for _ in range(num_trials)]
+
+        for idx in range(1, len(approximate_circuits)):
+            self.assertEqual(approximate_circuits[idx], approximate_circuits[0])
+
+    def test_gridsynth_unitary_deterministic(self):
+        """Test that calling gridsynth_unitary multiple times produces the same circuit."""
+        unitary = random_unitary(2, seed=12345)
+        num_trials = 10
+        approximate_circuits = [gridsynth_unitary(unitary.data, 1e-10) for _ in range(num_trials)]
+
+        for idx in range(1, len(approximate_circuits)):
+            self.assertEqual(approximate_circuits[idx], approximate_circuits[0])
+
+    @data(IGate(), XGate(), YGate(), ZGate(), HGate(), SGate(), SdgGate(), SXGate(), SXdgGate())
+    def test_clifford_matrix(self, clifford_gate):
```

Comment:
We typically try to avoid instantiations in the definitions and instead use a scheme, to push the work to when the test is actually executed
```
@data(IGate, XGate, ...)
def test_clifford_matrix(self, clifford_cls):
  clifford_gate = clifford_cls()
```